### PR TITLE
Fix Redis upstream healthcheck

### PR DESCRIPTION
**Pull request by vdatcu**
- This PR addresses the issue encountered in cloud environments where Redis is password protected
- The code has been refactored and commented
- Add global `logger` table with `error()`, `debug()` and `info()` functions which extend on the existing `ngx.log()`
- From the root dir, `./run_unit_tests.sh`, `make test-docker` and `make-test-docker-with-password`

### CODE REVIEW

File: src/lua/api-gateway/util/logger.lua [hunk 1]
```diff
@@ -67,6 +67,23 @@ local function _decorateLogger()
     end
 end
 
+logger = {}
+
+function logger.error(...)
+    _decorateLogger()
```
Comment (atrifan) [inline src/lua/api-gateway/util/logger.lua]: no need to decorage logger every time it's done once on _validateRequest to ensure that it is decorated across the request.

File: src/lua/api-gateway/util/logger.lua [hunk 2]
```diff
@@ -67,6 +67,23 @@ local function _decorateLogger()
...
+
+function logger.debug(...)
+    _decorateLogger()
+    ngx.log(ngx.DEBUG, ...)
+end
+
+function logger.info(...)
+    _decorateLogger()
+    ngx.log(ngx.INFO, ...)
+end
```
Comment (atrifan) [inline src/lua/api-gateway/util/logger.lua]: why not expose all the functions in the return bellow. and in factory require it above all other modules globally like so: logger = require ("....logger")

File: src/lua/api-gateway/redis/redisHealthCheck.lua [hunk 3]
```diff
@@ -27,24 +27,24 @@
...
-
-local base = require "api-gateway.validation.base"
-
-local HealthCheck = {}
+local RedisHealthCheck = {}
 local DEFAULT_SHARED_DICT = "cachedkeys"
+local HEALTHY_REDIS_UPSTREAM_KEY_PREFIX = "healthy_redis_upstream:"
 
-function HealthCheck:new(o)
+function RedisHealthCheck:new(o)
```
Comment (atrifan) [inline src/lua/api-gateway/redis/redisHealthCheck.lua:29]: is there any possibility that we can an add an UT for this module?

**Review by atrifan — CHANGES_REQUESTED** [review]

File: src/lua/api-gateway/redis/redisHealthCheck.lua (hunk 3 above)
Comment (vdatcu) [reply src/lua/api-gateway/redis/redisHealthCheck.lua:29]: Yes, will do now.

**Review by atrifan — APPROVED** [review]

**Review by atrifan — CHANGES_REQUESTED** [review]
running tests

**Review by atrifan — APPROVED** [review]

File: src/lua/api-gateway/redis/redisHealthCheck.lua [hunk 4]
```diff
@@ -19,158 +19,247 @@
...
+-- @param upstreamRedis Upstream address as host:port
+-- @return Upstream host
+-- @return Upstream port
 local function getHostAndPortInUpstream(upstreamRedis)
-    local p = {}
-    p.host = upstreamRedis
-
+    local upstreamAddress = {}
+    upstreamAddress.host = upstreamRedis
     local idx = string.find(upstreamRedis, ":", 1, true)
```
Comment (corinapurcarea) [inline src/lua/api-gateway/redis/redisHealthCheck.lua]: This logic can be replaced by a call to split function

**Review by corinapurcarea — CHANGES_REQUESTED** [review]

File: src/lua/api-gateway/redis/redisConnectionProvider.lua [hunk 5]
```diff
@@ -101,12 +101,19 @@ function RedisConnectionProvider:connectToRedis(host, port, password, redisTimeo
...
-
     if not ok then
         ngx.log(ngx.ERR, "Failed to connect to Redis instance: " .. host .. ", port: " .. port .. ". Error: ", err)
         return false, nil
     end
 
+    -- Check for existing connection
+    local times, error = redis:get_reused_times()
+
+    if times and times > 0 then
```
Comment (corinapurcarea) [inline src/lua/api-gateway/redis/redisConnectionProvider.lua]: I think you should check that times is not 0 ->
"If the connection comes from the connection pool, then the return value is always non-zero. "

**Review by corinapurcarea — CHANGES_REQUESTED** [review]

**Review by corinapurcarea — APPROVED** [review]
👍

**Pull request merged**